db/map.go

```diff
@@ -0,0 +1,319 @@
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Insert(list...)
+}
+
+func (tx WrappedTransaction) Update(list ...interface{}) (int64, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Update(list...)
+}
+
+func (tx WrappedTransaction) Delete(list ...interface{}) (int64, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Delete(list...)
+}
+
+func (tx WrappedTransaction) Select(holder interface{}, query string, args ...interface{}) ([]interface{}, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Select(holder, query, args...)
+}
+
+func (tx WrappedTransaction) SelectOne(holder interface{}, query string, args ...interface{}) error {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).SelectOne(holder, query, args...)
+}
+
+func (tx WrappedTransaction) Exec(query string, args ...interface{}) (sql.Result, error) {
+	return (WrappedExecutor{SqlExecutor: tx.Transaction}).Exec(query, args...)
+}
+
+var (
+	// selectTableRegexp matches the table name from an SQL select statement
+	selectTableRegexp = regexp.MustCompile(`(?i)^\s*select\s+[a-z\d\.\(\), \_\*` + "`" + `]+\s+from\s+([a-z\d\_,` + "`" + `]+)`)
+	// insertTableRegexp matches the table name from an SQL insert statement
+	insertTableRegexp = regexp.MustCompile(`(?i)^\s*insert\s+into\s+([a-z\d \_,` + "`" + `]+)\s+(?:set|\()`)
+	// updateTableRegexp matches the table name from an SQL update statement
+	updateTableRegexp = regexp.MustCompile(`(?i)^\s*update\s+([a-z\d \_,` + "`" + `]+)\s+set`)
+	// deleteTableRegexp matches the table name from an SQL delete statement
+	deleteTableRegexp = regexp.MustCompile(`(?i)^\s*delete\s+from\s+([a-z\d \_,` + "`" + `]+)\s+where`)
+
+	// tableRegexps is a list of regexps that tableFromQuery will try to use in
+	// succession to find the table name for an SQL query.
+	tableRegexps = []*regexp.Regexp{
+		selectTableRegexp,
+		insertTableRegexp,
+		updateTableRegexp,
+		deleteTableRegexp,
+	}
+)
+
+// tableFromQuery uses the tableRegexps on the provided query to return the
+// associated table name or an empty string if it can't be determined from the
+// query.
+func tableFromQuery(query string) string {
+	for _, r := range tableRegexps {
+		if matches := r.FindStringSubmatch(query); len(matches) >= 2 {
+			return matches[1]
+		}
+	}
+	return ""
+}
+
+// WrappedExecutor wraps a gorp.SqlExecutor such that its major functions
+// wrap error results in ErrDatabaseOp instances before returning them to the
+// caller.
+type WrappedExecutor struct {
+	gorp.SqlExecutor
+}
+
+func (we WrappedExecutor) Get(holder interface{}, keys ...interface{}) (interface{}, error) {
+	res, err := we.SqlExecutor.Get(holder, keys...)
+	if err != nil {
+		return res, ErrDatabaseOp{
+			Op:    "get",
+			Table: fmt.Sprintf("%T", holder),
+			Err:   err,
+		}
+	}
+	return res, err
+}
+
+func (we WrappedExecutor) Insert(list ...interface{}) error {
+	if err := we.SqlExecutor.Insert(list...); err != nil {
+		table := "unknown"
```

Comment:
Instead of deriving the table from the type, we could alter the `initTables` method to record the actual type <-> table mapping somewhere and use that here.

Reply:
This is a fairly big shift from the existing code. I'd rather take it as a follow-up if folks feel strongly. It does sound like it may be slightly better UX.

Reply:
:+1: works as a follow-up for me.

Reply:
Filed https://github.com/letsencrypt/boulder/issues/4594 so it doesn't get forgotten. Thanks!